src/stream/src/executor/source/mod.rs

```diff
@@ -60,10 +60,22 @@ pub fn get_split_offset_mapping_from_chunk(
     offset_idx: usize,
 ) -> Option<HashMap<SplitId, String>> {
     let mut split_offset_mapping = HashMap::new();
-    for (_, row) in chunk.rows() {
-        let split_id = row.datum_at(split_idx).unwrap().into_utf8().into();
-        let offset = row.datum_at(offset_idx).unwrap().into_utf8();
-        split_offset_mapping.insert(split_id, offset.to_string());
+    if chunk.cardinality() == 0 {
+        // assumes the chunk is a heartbeat chunk
```

Comment:
It appears that the current approach is less type-safe than the original `ChunkWithState` approach. Essentially we are mixing another kind of message into the output type (`StreamChunk`) of `into_chunk_stream`, so the branching here looks fragile to me. cc @xxchan What's your idea on this?

Reply:
Haven't fully checked it yet. It seems the old approach is equally unsafe as current approach? Both rely on invisible rows 🤡

Reply:
This is the best solution I can come up after the refactor of #14524

Reply:
Hi, I have resolved the above comments, any more concern about the "invisible chunk" approach? cc @BugenZhao @xxchan

Reply:
Is `StreamChunkWithState` or the heartbeat chunk here only passed from connector to SourceExecutor? If so, is it possible to pass a `Either<StreamChunk, SourceOffset>` instead?

Btw, substituting `StreamChunk with offset column` for `StreamChunkWithState` (https://github.com/risingwavelabs/risingwave/pull/14524) seems to be a orthogonol problem to me, because neither of them can pass source offset when no input i.e. heartbeat.

Reply:
>  If so, is it possible to pass a Either<StreamChunk, SourceOffset> instead?

I am confused. In current framework (main), offset of source message has been parsed into a column in the `StreamChunk`. So do you mean we need to add a new type (`SourceOffset`) to the `Message`?
https://github.com/risingwavelabs/risingwave/blob/7ce1b6aadef6061b2f53939106d937c50963a8af/src/stream/src/executor/mod.rs#L796-L800

That way we also need to refactor other part to support adding a new message type.

Reply:
What looks weird to me is that we have a specific path for the chunk with cardinality of zero. What about defining the convention like this:

- All visible rows will be forwarded to the downstream.
- All rows (including those visible or invisible) will be used to update the offset.

The behavior will be similar to what we implement in this PR, but does not require the ad-hoc path and is more clearly defined.

Reply:
> All rows (including those visible or invisible) will be used to update the offset.

+1. I refactored the code.  Is it looks good to you? @tabVersion

Reply:
> > I am confused. In current framework (main), offset of source message has been parsed into a column in the StreamChunk. So do you mean we need to add a new type (SourceOffset) to the Message?
>
> What looks weird to me is that we have a specific path for the chunk with cardinality of zero.

You guys are right. That will be 2 code paths, which looks even worse.